test/distributed/fsdp/test_fsdp_grad_acc.py

```diff
@@ -0,0 +1,239 @@
+                loss.backward()
+                losses.append(loss)
+        output = fsdp_model(*batches[-1])
+        loss = fsdp_model.module.get_loss(batches[-1], output)
+        loss.backward()
+        losses.append(loss)
+        acc_loss = sum(losses)
+        acc_grads = [p.grad.detach().clone() for p in fsdp_model.parameters()]
+
+        # Compare the losses and gradients
+        torch.testing.assert_allclose(ref_loss, acc_loss)
+        assert len(ref_grads) == len(acc_grads)
+        for ref_grad, acc_grad in zip(ref_grads, acc_grads):
+            assert ref_grad.device == acc_grad.device
+            assert ref_grad.size() == acc_grad.size()
+            assert ref_grad.dtype == acc_grad.dtype
+            torch.testing.assert_allclose(ref_grad, acc_grad)
+
+        # Restore the default dtype
+        torch.set_default_dtype(old_default_dtype)
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize(
+        "num_iters_to_acc",
+        [2, 3, 4],
+    )
+    def test_no_sync(self, num_iters_to_acc):
+        """Tests the ``no_sync()`` context manager."""
+        assert num_iters_to_acc >= 2, \
+            "Accumulate for at least 2 iterations to be nontrivial"
+        self._test_no_sync(batch_dim=1, num_iters_to_acc=num_iters_to_acc)
+
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize(
+        "nested_model",
+        [False, True],
+    )
+    @parametrize(
+        "use_no_sync",
+        [False, True],
+    )
+    def test_no_sync_communication(self, nested_model: bool, use_no_sync: bool):
+        """
+        Tests the ``no_sync()`` context manager by checking the communication
+        cost in terms of calls to collective communication primitives.
+
+        Arguments:
+            nested_model (bool): If ``True``, uses ``NestedWrappedModule``,
+                which has nested FSDP instances; if ``False``, uses the default
+                model, which does not have nested FSDP instances.
+            use_no_sync (bool): If ``True``, uses the ``no_sync()`` context
+                manager to accumulate gradients for one iteration before
+                synchronizing gradients in the second iteration; if ``False``,
+                only checks the communication cost of a normal iteration.
+        """
+        group = dist.distributed_c10d._get_default_group()
+        device = torch.device("cuda")
+        if nested_model:
+            model = NestedWrappedModule(group, wrap_fsdp=True)
+            fsdp_model: FSDP = FSDP(model, group).to(device)
+        else:
+            fsdp_model: FSDP = self._get_wrapped_model(
+                group, cuda_first=False,
+            )
+        batch = fsdp_model.module.get_input(device)
+
+        num_fsdp = 0
+        for m in fsdp_model.modules():  # includes `self`
+            if isinstance(m, FSDP) and len(m.params) > 0:
+                num_fsdp += 1
+
+        # Not all PyTorch versions have the `_base` methods available
```

Comment:
I'm not sure if we need this as we don't test against all PyTorch versions, only PT master which should have both of these. In PT FSDP codebase I believe we unconditionally use them too

Reply:
Yup, fixed.